testing/test_pdb.py

```diff
@@ -318,8 +342,11 @@ class Config(DefaultConfig):
         pygments.formatters.TerminalTrueColorFormatter
     )
 
-    assert Pdb(Config=Config).format_source("print(42)") == (
-        "\x1b[38;2;0;128;0;01mprint\x1b[39;00m(\x1b[38;2;102;102;102m42\x1b[39m)\n"
+    formatted_source = Pdb(Config=Config).format_source("print(42)")
+    assert formatted_source in (
+        '\x1b[38;2;0;128;0;01mprint\x1b[39;00m(\x1b[38;2;102;102;102m42\x1b[39m)\n',
+        # Might happen on AppVeyor?
+        '\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n',
```

Comment:
@dougthor42 
Any idea about this?  It started already when I've started messing with it, so might be due to external deps?

Reply:
## TL;DR

I have no idea. I cannot get anything to return the value `\x1b[38;2;0;128;0;01m` (which includes the `;01`).

For me on both Windows and Linux (WSL):

```python
# This assert fails
assert formatted_source == '\x1b[38;2;0;128;0;01mprint\x1b[39;00m(\x1b[38;2;102;102;102m42\x1b[39m)\n'

# This assert passes
assert formatted_source == '\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'
```


## Questions

### Question 1:

Do you know what the last `;01` in the first escape sequence is supposed to be? It looks like it's the [24-bit sequence](https://en.wikipedia.org/wiki/ANSI_escape_code#24-bit), but the Wikipedia paged doesn't list a 6th term.

```
24-bit
String:     \x1b [     38    ;  2 ;  0 ;  128 ;   0 ;  01  m
Translated:  ESC [ foreground; dim; red; green; blue; ???  m
                                                       ^
```


### Question 2:

Is there a particular reason why you wanted to change this test to use the full string? It looks like the previous iteration would pass both.

This is mostly curiosity. Using the full string is a more precise test (but more fragile as we see here).


### Question 3:

What do you get when you run this locally?

```python
>>> from pygments.formatters import TerminalTrueColorFormatter
>>> from pygments.lexers import PythonLexer
>>> from pygments import highlight
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
```

----

The rest of this post is just me going through ideas. It can probably be ignored.

## Investigations

### The Terminals

For Windows, I have CMD and PowerShell terminals available. For Linux, I have whatever WSL uses. It looks and acts like CMD, so I assume it's the same thing behind the scenes.

I checked all 3 possible combinations and

| Terminal | OS | without `;01` in the expected string | with `;01` in the expected string |
| --- | --- | :-: | :-: |
| CMD | Windows | Pass | Fail |
| CMD | Linux (WSL) | Pass | Fail |
| PowerShell | Windows | Pass | Fail |

If I just use pygments to format that string:

```python
>>> # Python 3.6.8 on win32+CMD, pygments==2.5.2
>>> from pygments.formatters import TerminalTrueColorFormatter
>>> from pygments.lexers import PythonLexer
>>> from pygments import highlight
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
'\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'

>>> # Python 3.6.8 on linux+CMD, pygments==2.5.2
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
'\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'

>>> # Python 3.6.8 on win32+PowerShell, pygments==2.5.2
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
'\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'
```

In short, I have no idea what's going on.


### The Lexers

Turns out this wasn't the root cause, so this can be ignored.

I thought perhaps that the tokens for that sting were different on linux and Windows. This is not the case (they're the same), but I did notice that things change with pygments versions. Specifically, the `print` string is a `Token.Keyword` in pygments 2.4.2 and is a `Token.Name.Builtin` in pygments 2.5.2.

The tokens generated by the string `print(42)` are:

```python
>>> # Python 3.6.8 on Linux, pygments==2.4.2
>>> from pygments.lexers import PythonLexer
>>> list(PythonLexer().get_tokens("print(42)"))
[(Token.Keyword, 'print'),
 (Token.Punctuation, '('),
 (Token.Literal.Number.Integer, '42'),
 (Token.Punctuation, ')'),
 (Token.Text, '\n')]
```

```python
>>> # Python 3.6.3 on win32, pygments==2.5.2
>>> from pygments.lexers import PythonLexer
>>> list(PythonLexer().get_tokens("print(42)"))
[(Token.Name.Builtin, 'print'),
 (Token.Punctuation, '('),
 (Token.Literal.Number.Integer, '42'),
 (Token.Punctuation, ')'),
 (Token.Text, '\n')]
```

The reason this happens is because in pygments 2.4.2, the `PythonLexer` was for Python2 and a specific `Python3Lexer` existed. In pygments > 2.5, these were switched. `PythonLexer` is for Python3 and a specific `Python2Lexer` exists.


### packages

I added `python -m pip list` to the tox commands and then ran the failing test on both Linux and Windows.

On Windows, the following packages are added:
+ atomicwrites (from pytest)
+ colorama (from pytest)
+ pyreadline (from pdbpp)

It's possible, but unlikely, that one of these is mucking things up. I updated the tox env to remove `colorama` and `pyreadline`, but that had no effect. I could not remove `atomicwrites` without breaking pytest, so that wasn't checked.

I don't think this is the issue.


### Something specific in pdbpp

L733 `format_source` is what actually adds the escape codes. It determines the highlighting function and calls that as the return value.

If we force that to use the pygments lexers and formatters directly rather than going through `_get_source_highlight_function`, the I still do not see the `;01` in the output.

```diff
         src = self.try_to_decode(src)
-        return self._highlight(src)
+        import pygments
+        return pygments.highlight(src,
+            pygments.lexers.PythonLexer(stripnl=False),
+            pygments.formatters.TerminalTrueColorFormatter()
+        )
```

Basically all this does is confirm that `_get_source_highlight_function` and `_get_pygments_formatter` doesn't have any strangeness in it.

Reply:
> I have no idea. I cannot get anything to return the value `\x1b[38;2;0;128;0;01m` (which includes the `;01`).

The ";01" means bold here and IIRC comes from a newer version of pygments, but have not really investigated myself.

See 3265207 (on master), where I've made the test less strict.

Reply:
> ";01" means bold here

I guess I figured that, but then second-guessed myself because the wiki page for 24-bit color doesn't mention using other modifiers.

At least for my terminal, there's no difference between with and without `;01`:

![image](https://user-images.githubusercontent.com/5386897/71595577-0fdf5180-2af1-11ea-8913-9e98334d24ed.png)


Ah well, no biggie. I'm fine with using the less-strict test.